kv/dist_sender.go

```diff
@@ -406,90 +379,141 @@ func (ds *DistSender) sendRPC(trace *tracer.Trace, rangeID proto.RangeID, replic
-	defer trace.Epoch("sending RPC")()
-	// Truncate the request to our current range, making sure not to
-	// touch it unless we have to.
-	if _, isReverseScan := args.(*proto.ReverseScanRequest); isReverseScan {
-		// Truncate to the current range if the reverse scan crosses range boundaries.
-		if key := args.Header().Key; key != nil && key.Less(desc.StartKey) {
-			defer func(k proto.Key) { args.Header().Key = k }(key)
-			args.Header().Key = desc.StartKey
+// truncate restricts all contained requests to the given key range.
+// Even on error, the returned closure must be executed; it undoes any
+// truncations performed.
+// First, the boundaries of the truncation are obtained: This is the
+// intersection between [from,to) and the descriptor's range.
+// Secondly, all requests contained in the batch are "truncated" to
+// the resulting range, inserting NoopRequest appropriately to
+// replace requests which are left without a key range to operate on.
+// The number of non-noop requests after truncation is returned along
+// with a closure which must be executed to undo the truncation, even
+// in case of an error.
+func truncate(br *proto.BatchRequest, desc *proto.RangeDescriptor, from, to proto.Key) (func(), int, error) {
+	if !desc.ContainsKey(keys.KeyAddress(from)) {
+		from = desc.StartKey
+	}
+	if !desc.ContainsKeyRange(desc.StartKey, keys.KeyAddress(to)) || to == nil {
+		to = desc.EndKey
+	}
+	fromAddr := keys.KeyAddress(from)
+	toAddr := keys.KeyAddress(to)
+	truncateOne := func(args proto.Request) (bool, []func(), error) {
+		header := args.Header()
+		if !proto.IsRange(args) {
+			if len(header.EndKey) > 0 {
+				return false, nil, util.Errorf("%T is not a range command, but EndKey is set", args)
+			}
+			if !desc.ContainsKey(keys.KeyAddress(args.Header().Key)) {
+				return true, nil, nil
+			}
+			return false, nil, nil
 		}
-	} else {
-		// It is illegal to send EndKey on commands which do not operate on ranges.
-		if endKey := args.Header().EndKey; endKey != nil && !endKey.Less(desc.EndKey) {
-			defer func(k proto.Key) { args.Header().EndKey = k }(endKey)
-			args.Header().EndKey = desc.EndKey
+		var undo []func()
+		key, endKey := args.Header().Key, args.Header().EndKey
+		keyAddr, endKeyAddr := keys.KeyAddress(key), keys.KeyAddress(endKey)
+		if keyAddr.Less(fromAddr) {
+			undo = append(undo, func() { args.Header().Key = key })
+			args.Header().Key = from
+			keyAddr = fromAddr
+		}
+		if !endKeyAddr.Less(toAddr) {
+			undo = append(undo, func() { args.Header().EndKey = endKey })
+			args.Header().EndKey = to
+			endKeyAddr = toAddr
+		}
+		// Check whether the truncation has left any keys in the range. If not,
+		// we need to cut it out of the request.
+		return !keyAddr.Less(endKeyAddr), undo, nil
+	}
+
+	var fns []func()
+	gUndo := func() {
+		for _, f := range fns {
+			f()
+		}
+	}
+
+	var numNoop int
+	for pos, arg := range br.Requests {
```

Comment:
Honestly this is just home-grown out of the previous code which also did this temporary truncation. It could as well return a new batch request. Seeing that the majority of requests will never need truncation (hopefully), that could be ill-advised performance wise, but I have no indication to believe that at some point. I'll (re-)add a TODO/comment I had about that in some of the intermediate commits back in.